File: Stack/MinStack.py
# 155. 最小栈
# 设计一个支持 push ，pop ，top 操作，并能在常数时间内检索到最小元素的栈。
#
# push(x) —— 将元素 x 推入栈中。
# pop() —— 删除栈顶的元素。
# top() —— 获取栈顶元素。
# getMin() —— 检索栈中的最小元素。


#辅助栈算法
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack=list()
        self.helpStack=[2**31-1,]


    def push(self, x: int) -> None:
        self.stack.append(x)
        if x<self.helpStack[-1]:
            self.helpStack.append(x)
        else:
            self.helpStack.append(self.helpStack[-1])

    def pop(self) -> None:
        self.stack.pop()
        self.helpStack.pop()


    def top(self) -> int:
        return self.stack[-1]


    def getMin(self) -> int:
        return self.helpStack[-1]

File: data_structure/stack_op/min_stack.go
package stack_op

/*
剑指 Offer 30. 包含min函数的栈
定义栈的数据结构，请在该类型中实现一个能够得到栈的最小元素的 min 函数在该栈中，调用 min、push 及 pop 的时间复杂度都是 O(1)。
*/

type MinStack struct {
	arr []int
	min int
}

/** initialize your data structure here. */
func Constructor1() MinStack {
	return MinStack{
		arr: []int{},
		min: 0,
	}
}

func (this *MinStack) Push(x int) {
	if len(this.arr) == 0 {
		this.min = x
	}
	diff := x - this.min
	if diff < 0 {
		this.min = x
	}
	this.arr = append(this.arr, diff)
}

func (this *MinStack) Pop() {
	if len(this.arr) == 0 {
		return
	}
	topDiff := this.arr[len(this.arr)-1]
	if topDiff < 0 {
		this.min = this.min - topDiff
	}
	this.arr = this.arr[:len(this.arr)-1]
}

func (this *MinStack) Top() int {
	if len(this.arr) == 0 {
		return -1
	}
	topDiff := this.arr[len(this.arr)-1]
	if topDiff > 0 {
		return this.min + topDiff
	}
	return this.min
}

func (this *MinStack) Min() int {
	return this.min
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.Min();
 */
